pinot-spi/src/main/java/org/apache/pinot/spi/trace/Tracer.java

```diff
@@ -0,0 +1,35 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.spi.trace;
+
+public interface Tracer {
```

Comment:
So to understand the thread model here --- each query worker (could also be a runner ?) thread uses the Tracer to register the current query's requestId and start an invocation span for a particular operator ?

Since each worker thread executes multiple operator trees (1 per segment), in theory the implementor (which requires operator code changes) has the freedom to collect InvocationRecording (s) with span starting and ending after each segment. Is this understanding correct ?

Reply:
The requestId gets registered at the same point it always did, and then it gets propagated however the implementation wants. The SPI doesn't model this propagation because it would lead to debate about which fields should be propagated and it's best handled by the implementation. But once the requestId is attached to a thread, work is associated with the request. 

This strays into the implementation details of my plugin, but demonstrates that it is possible to differentiate
between segments with these primitives. On the query worker threads, there is a stack which recordings are pushed onto and a thread local counter which identify recordings within the context of the request. Since invocations are synchronous, the stack is empty when work starts on a segment, and becomes empty after the segment is processed. Each recording gets a number assigned to it. When processing the next segment starts, the stack will be empty again, but the recording counter does not get reset until there is a new request, so the recordings for different segments are distinct and can be differentiated by their numbers.

It would be nice to capture the segment names though.